--- ./smallapp/unbound-checkconf.c	2017-05-16 14:39:24.000000000 +0200
+++ ./smallapp/unbound-checkconf.c.new	2017-07-02 13:39:49.332042712 +0200
@@ -426,10 +426,12 @@
 		cfg->trusted_keys_file_list, cfg->chrootdir, cfg);
 	check_chroot_string("dlv-anchor-file", &cfg->dlv_anchor_file,
 		cfg->chrootdir, cfg);
+#if 0
 #ifdef USE_IPSECMOD
 	check_chroot_string("ipsecmod-hook", &cfg->ipsecmod_hook, cfg->chrootdir,
 		cfg);
 #endif
+#endif
 	/* remove chroot setting so that modules are not stripping pathnames*/
 	free(cfg->chrootdir);
 	cfg->chrootdir = NULL;
