diff -up unbound-1.1.0/daemon/daemon.c.log_open unbound-1.1.0/daemon/daemon.c
--- unbound-1.1.0/daemon/daemon.c.log_open	2008-11-19 15:36:53.000000000 +0100
+++ unbound-1.1.0/daemon/daemon.c	2008-11-19 15:37:34.000000000 +0100
@@ -417,7 +417,6 @@ daemon_fork(struct daemon* daemon)
 	signal_handling_playback(daemon->workers[0]);
 
 	/* Start resolver service on main thread. */
-	log_info("start of service (%s).", PACKAGE_STRING);
 	worker_work(daemon->workers[0]);
 	log_info("service stopped (%s).", PACKAGE_STRING);
 
diff -up unbound-1.1.0/daemon/unbound.c.log_open unbound-1.1.0/daemon/unbound.c
--- unbound-1.1.0/daemon/unbound.c.log_open	2008-11-19 15:38:49.000000000 +0100
+++ unbound-1.1.0/daemon/unbound.c	2008-11-19 15:39:11.000000000 +0100
@@ -392,6 +392,8 @@ perform_setup(struct daemon* daemon, str
 	(void)daemon;
 #endif
 
+	log_info("start of service (%s).", PACKAGE_STRING);
+
 	/* box into the chroot */
 #ifdef HAVE_CHROOT
 	if(cfg->chrootdir && cfg->chrootdir[0]) {
