diff -Naur unbound-1.5.9-orig/iterator/iterator.c unbound-1.5.9/iterator/iterator.c
--- unbound-1.5.9-orig/iterator/iterator.c	2016-05-31 12:55:22.000000000 -0400
+++ unbound-1.5.9/iterator/iterator.c	2016-06-16 09:11:13.510969049 -0400
@@ -2024,6 +2024,8 @@
 		iq->qinfo_out.qname_len = iq->qchase.qname_len;
 		iq->minimise_count++;
 
+		iter_dec_attempts(iq->dp, 1);
+
 		/* Limit number of iterations for QNAMEs with more
 		 * than MAX_MINIMISE_COUNT labels. Send first MINIMISE_ONE_LAB
 		 * labels of QNAME always individually.
